trin-core/src/portalnet/overlay_service.rs

```diff
@@ -827,23 +827,116 @@ impl OverlayService {
-    fn process_content(&mut self, _content: Content, source: Enr) {
+    fn process_content(&mut self, content: Content, source: Enr) {
         debug!(
             "[{:?}] Processing Content response from node. Node: {}",
             self.protocol,
             source.node_id()
         );
-        // TODO: Implement processing logic.
+        if let Content::Enrs(enrs) = content {
+            self.process_discovered_enrs(enrs.into_iter().map(|ssz_enr| ssz_enr.into()).collect());
+        }
+    }
+
+    /// Processes a collection of discovered nodes.
+    fn process_discovered_enrs(&mut self, enrs: Vec<Enr>) {
+        let local_node_id = self.local_enr().node_id();
+
+        // Acquire write lock here so that we can perform node lookup and insert/update atomically.
+        // Once we acquire the write lock for the routing table, there are no other locks that we
+        // need to acquire, so we should not create a deadlock.
+        let mut kbuckets = self.kbuckets.write();
+
+        for enr in enrs {
+            let node_id = enr.node_id();
+
+            // Ignore ourself.
+            if node_id == local_node_id {
+                continue;
+            }
+
+            let key = kbucket::Key::from(node_id);
+            let node = match kbuckets.entry(&key) {
+                kbucket::Entry::Present(entry, _) => Some(entry.value().clone()),
+                kbucket::Entry::Pending(ref mut entry, _) => Some(entry.value().clone()),
+                _ => None,
+            };
+
+            // If the node is in the routing table, then check to see if we should update its entry.
+            // If the node is not in the routing table, then add the node in a disconnected state.
+            // A subsequent ping will establish connectivity with the node. If the insertion succeeds,
+            // then add the node to the ping queue. Ignore insertion failures.
+            if let Some(present_node) = node {
+                if present_node.enr().seq() < enr.seq() {
+                    let updated_node = Node {
+                        enr,
+                        data_radius: present_node.data_radius(),
+                    };
+
+                    // The update removed the node because it would violate the incoming peers condition
+                    // or a bucket/table filter. Remove the node from the ping queue.
+                    if let UpdateResult::Failed(reason) =
+                        kbuckets.update_node(&key, updated_node, None)
+                    {
+                        self.peers_to_ping.remove(&node_id);
+                        debug!(
+                            "Failed to update discovered node. Node: {}, Reason: {:?}",
+                            node_id, reason
+                        );
+                    }
+                }
+            } else {
+                let node = Node {
+                    enr,
+                    data_radius: U256::from(u64::MAX),
```

Comment:
I think this is for another PR but it is probably worth adding a data structure for the radius, something like `DataRadius`  and implementing Default trait, so this line will look something like:

```rust
data_radius: DataRadius::default()
```

Reply:
I was thinking a constant, but like you said, I think this is best left for a later PR.